### PR TITLE
feat(dot/parachain/backing): handle `second` incoming overseer message

### LINKED ISSUE #3506: feat(dot/parachain/backing): handle `Second` incoming overseer message
## Issue summary

- Can only be tackled after https://github.com/ChainSafe/gossamer/pull/3497 is merged
- Here in Gossamer 
https://github.com/ChainSafe/gossamer/blob/8f10724821e8d3a4ae502e0c5e8691bd98cdd12f/dot/parachain/backing/candidate_backing.go#L109-L110

## Other information and links

Reference code in polkadot-sdk
https://github.com/paritytech/polkadot-sdk/blob/769bdd3ff33a291cbc70a800a3830638467e42a2/polkadot/node/core/backing/src/lib.rs#L748-L749

### PR DESCRIPTION
**Pull request by axaysagathiya**
## Note
need to merge ONLY AFTER #3600 gets merged

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

closes #3506 

## Primary Reviewer

@kishansagathiya @EclesioMeloJunior

### CODE REVIEW

File: dot/parachain/backing/candidate_backing.go
Comment (kanishkatn) [inline dot/parachain/backing/candidate_backing.go:232]: Could you tell me what this check is for? 

Also, can we move this to a function?

File: dot/parachain/backing/second.go
Comment (kanishkatn) [inline dot/parachain/backing/second.go:22]: nit:

Is this error message copied from polkadot?

File: dot/parachain/backing/candidate_backing.go
Comment (EclesioMeloJunior) [reply dot/parachain/backing/candidate_backing.go:232]: Also you can use a switch:

```go
switch err {
case errWrongPVDForSecondingCandidate, 
  errUnknownRelayParentForSecondingCandidate,
  ...:
    logger.Error(err.Error())
    return nil
}
```

Comment (kanishkatn) [reply dot/parachain/backing/candidate_backing.go:232]: We may not need a function if it's using switch!

**Review by kanishkatn — APPROVED** [review]
lgtm!

File: dot/parachain/backing/candidate_backing.go
Comment (axaysagathiya) [reply dot/parachain/backing/candidate_backing.go:232]: > Could you tell me what this check is for?

In Rust, they are sometimes logging errors and returning the OK value instead of an error.
But I have returned an error from the function, and I am making sure we return nil error further for these particular errors.

File: dot/parachain/backing/second.go
Comment (axaysagathiya) [reply dot/parachain/backing/second.go:22]: Yes, the error messages are copied from Polkadot.

Comment (kanishkatn) [reply dot/parachain/backing/second.go:22]: I see. I kinda look at the node as a singular entity, so I'd suggest to remove "we". 

I'm just curious as to what the rest of the team thinks and feel free to resolve this before merging if you're in a hurry!

File: dot/parachain/backing/second.go
Comment (EclesioMeloJunior) [inline dot/parachain/backing/second.go:42]: I would say that given the error context can be useful for log informations, for example in this case if `A != B` then we just return `errWrongValue` but will be better if we return `errWrongValue: expected A got B`

```suggestion
		return fmt.Errorf("%w: mismatch between %v and %v", errWrongPVDForSecondingCandidate, 
		  candidateReceipt.Descriptor.PersistedValidationDataHash,
		  pvdHash)
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/parachain/backing/second.go
Comment (kishansagathiya) [inline dot/parachain/backing/second.go:45]: I think we need to calculate Hash of PVD at a few places. We could add this logic as a method to `parachaintypes.PersistedValidationData` to be able to reuse this code.

File: dot/parachain/backing/second.go
Comment (kishansagathiya) [inline dot/parachain/backing/second.go:48]: This will just print two random hashes, which doesn't mean much to us. We could make this more informative like 
```suggestion
		return fmt.Errorf("%w: mismatch between persisted validation hash in candidate decriptor: %s and calculated hash of given persisted validation data: %s",
```

**Review by kishansagathiya — APPROVED** [review]
Will wait for the other PR to merge first, but this looks good to me!!
:tada:

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**
